packages/components/src/ui/context/polymorphic-component.ts

```diff
@@ -11,24 +11,30 @@ import type * as React from 'react';
  * by `ComponentPropsWithRef`. The context is that components should require the `children`
  * prop explicitely when needed (see https://github.com/WordPress/gutenberg/pull/31817).
  */
-export type PolymorphicComponentProps< P, T extends React.ElementType > = P &
-	Omit< React.ComponentPropsWithRef< T >, 'as' | keyof P | 'children' > & {
-		as?: T | keyof JSX.IntrinsicElements;
-	};
+export type PolymorphicComponentProps<
+	P,
+	T extends React.ElementType,
+	IsPolymorphic extends boolean = true
+> = P &
+	Omit< React.ComponentPropsWithRef< T >, 'as' | keyof P | 'children' > &
+	( IsPolymorphic extends true
+		? {
+				as?: T | keyof JSX.IntrinsicElements;
+		  }
+		: { as?: never } );
 
-export type ElementTypeFromPolymorphicComponentProps<
-	P
-> = P extends PolymorphicComponentProps< unknown, infer T > ? T : never;
-
-export type PropsFromPolymorphicComponentProps<
-	P
-> = P extends PolymorphicComponentProps< infer PP, any > ? PP : never;
-
-export type PolymorphicComponent< T extends React.ElementType, O > = {
+export type PolymorphicComponent<
+	T extends React.ElementType,
+	O,
+	IsPolymorphic extends boolean
```

Comment:
Why doesn't `IsPolymorphic` here default to `true`, like it does in `PolymorphicComponentProps` ?

Reply:
Just because PolymorphicComponent is never used directly, so there's no reason to give it a default. It helps with compiler warnings when casting PolymorphicComponentProps to a PolymorphicComponent in case there's ever a mistake there.

Reply:
Got it, thank you for the explanation. Feel free to ignore this comment :)